Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch

```diff
@@ -0,0 +1,294 @@
++            buf.readBytes(signature);
++
++            byte[] data = new byte[buf.readableBytes()];
++            buf.getBytes(buf.readerIndex(), data);
++
++            try {
++                javax.crypto.Mac mac = javax.crypto.Mac.getInstance("HmacSHA256");
++                mac.init(new javax.crypto.spec.SecretKeySpec(com.destroystokyo.paper.PaperConfig.velocitySecretKey, "HmacSHA256"));
++                byte[] mySignature = mac.doFinal(data);
++                if (!java.security.MessageDigest.isEqual(signature, mySignature)) {
++                    this.disconnect("Unable to verify player details");
++                    return;
++                }
++            } catch (java.security.NoSuchAlgorithmException | java.security.InvalidKeyException e) {
++                throw new AssertionError(e);
++            }
++
++            // Populate the spoofed game profile
++            this.networkManager.l = new java.net.InetSocketAddress(com.google.common.net.InetAddresses.forString(buf.readUtf(32767)), ((java.net.InetSocketAddress) this.networkManager.getSocketAddress()).getPort());
++            GameProfile profile = new GameProfile(buf.readUuid(), buf.readUtf(16));
++            int propertiesToRead = buf.readVarInt();
++            for (int i1 = 0; i1 < propertiesToRead; i1++) {
++                String propertyName = buf.readUtf(32767);
++                String propertyValue = buf.readUtf(32767);
++                String propertySignature = buf.readBoolean() ? buf.readUtf(32767) : null;
++                profile.getProperties().put(propertyName, new com.mojang.authlib.properties.Property(propertyName, propertyValue, propertySignature));
++            }
++            this.setGameProfile(profile);
++
++            // Proceed with login
++            authenticatorPool.execute(() -> {
++                try {
++                    new LoginHandler().fireEvents();
++                } catch (Exception ex) {
++                    disconnect("Failed to verify username!");
++                    server.server.getLogger().log(java.util.logging.Level.WARNING, "Exception verifying " + i.getName(), ex);
++                }
++            });
++            return;
++        }
++        // Paper end
+         this.disconnect(new ChatMessage("multiplayer.disconnect.unexpected_query_response", new Object[0]));
+     }
+ 
+diff --git a/src/main/java/net/minecraft/server/NameReferencingFileConverter.java b/src/main/java/net/minecraft/server/NameReferencingFileConverter.java
+index 8a2ff6a41..d9839cb48 100644
+--- a/src/main/java/net/minecraft/server/NameReferencingFileConverter.java
++++ b/src/main/java/net/minecraft/server/NameReferencingFileConverter.java
+@@ -61,6 +61,7 @@ public class NameReferencingFileConverter {
+         });
+ 
+         if (minecraftserver.getOnlineMode()
++                || (com.destroystokyo.paper.PaperConfig.velocitySupport && com.destroystokyo.paper.PaperConfig.velocityOnlineMode) // Paper - Velocity support
```

Comment:
We really need to abstract all this now to a single method PaperConfig.isProxyOnlineMode()

Reply:
I was considering that, but wondered if we wanted a separate patch for that.

Reply:
We can put in the stub method in the bungee patch, then velocity amends that.

Reply:
Will do this tonight.